### PR TITLE
File Importer Error UI

### LINKED ISSUE #798: Admin can download failed rows from import
Part of epic #528

**What type of user is this for? [admin/supervisor/volunteer/all]**
admin

**Where does/should this occur?**
On the imports page accessible via the Admin dashboard. Click the drop down menu in the upper right hand corner of header and select "System Imports".

<a href="https://casa-qa.herokuapp.com/">Log in to QA</a> and use these credentials to access the admin dashboard:
username – casa_admin1@example.com
password – 123456

**Description**
After user imports a file, any failed rows should be available to download immediately after the import is complete. The completion of an import should trigger a notification on the page that reads:

"Not all rows were imported. Click here to download failed rows." << The second sentence should be hyperlinked, and launch the download of a .csv containing failed rows.

This resolves and replaces #707

### PR DESCRIPTION
**Pull request by ds29**
### What github issue is this PR for, if any?
Resolves #798 and #636 

### What changed, and why?

- Refactored the behavior of importers to handle error tracking and enable the ability to export (download) failed rows
  - `UserImporter` was split into `VolunteerImporter` and `SupervisorImporter` for smaller class size and better clarity
  - Updated related references and specs
- Added a new route (action) for `download_failed` to `ImportsController` enabling download
- Updated the importer tab views to set the `download` attribute signaling browsers to download the file instead of navigating to it as a normal page.
- Introduced `session` storage to allow the failed rows to be downloaded on request to eliminate any need for file storage or management.
- Introduced a modal to handle error dialogues
- Refactored general importer UI
- Added CSV header validation
- Stopped using flash for error messages
- Submit buttons are disabled when clicked
- Submit buttons are disabled when no file is selected

### How will this affect user permissions?
- Volunteer permissions: No change
- Supervisor permissions: No change
- Admin permissions: No change

### How is this tested? (please write tests!) 💖💪

This is tested by updates to existing test coverage for the imports.

### Screenshots please :)
Screen Shot 2020-09-23 at 10 54 49 PM
Screen Shot 2020-09-23 at 10 53 43 PM
Screen Shot 2020-09-23 at 10 52 54 PM
Screen Shot 2020-09-23 at 10 52 35 PM
Screen Shot 2020-09-23 at 10 52 07 PM
Screen Shot 2020-09-23 at 10 51 56 PM
Screen Shot 2020-09-23 at 10 51 38 PM

### Feelings gif (optional)
Spreadsheets

### CODE REVIEW

**Comment by ryanmrodriguez** [comment]
Resolves https://github.com/rubyforgood/casa/issues/636. As well as general code cleanup and UI refactoring.

File: app/views/imports/_cases.html.erb [hunk 1]
```diff
@@ -1,20 +1,37 @@
...
   </ul>
-  <%= f.file_field :file, accept: 'text/csv', class: 'form-control-file', style: "margin: auto;" %>
-  <br>
-  <%= button_tag :class => "btn btn-md btn-success" do %>
-    <i class="fa fa-upload"></i> Import Cases CSV
-  <% end %>
+  <%= f.file_field :file, id: 'case-file', accept: 'text/csv', class: 'form-control-file', style: "margin: auto;" %>
+</div>
+
+<%= button_tag id: "case-import-button", :class => "btn btn-md btn-success pull-right",
```
Comment (KatherineMuedas) [inline app/views/imports/_cases.html.erb:26]: This is more like a personal preference so it is not a blocker, I think we could change the `:class => "btn btn-md btn-success pull-right"` to use the new format `class: "btn btn-md btn-success pull-right"`

File: app/views/imports/_supervisors.html.erb [hunk 2]
```diff
@@ -1,22 +1,36 @@
...
   </ul>
-  <%= f.file_field :file, accept: 'text/csv', class: 'form-control-file', style: "margin: auto;" %>
-  <br>
-  <%= button_tag :class => "btn btn-md btn-success" do %>
-    <i class="fa fa-upload"></i> Import Supervisors CSV
-  <% end %>
+  <%= f.file_field :file, id: 'supervisor-file', accept: 'text/csv', class: 'form-control-file', style: "margin: auto;" %>
+</div>
+
+<%= button_tag id: "supervisor-import-button", :class => "btn btn-md btn-success pull-right",
```
Comment (KatherineMuedas) [inline app/views/imports/_supervisors.html.erb:26]: same here `:class =>` to  `class: `

File: app/views/imports/_volunteers.html.erb [hunk 3]
```diff
@@ -1,22 +1,36 @@
...
   </ul>
-  <%= f.file_field :file, accept: 'text/csv', class: 'form-control-file', style: "margin: auto;" %>
-  <br>
-  <%= button_tag :class => "btn btn-md btn-success" do %>
-    <i class="fa fa-upload"></i> Import Volunteers CSV
-  <% end %>
+  <%= f.file_field :file, id: 'volunteer-file', accept: 'text/csv', class: 'form-control-file', style: "margin: auto;" %>
+</div>
+
+<%= button_tag id: "volunteer-import-button", :class => "btn btn-md btn-success pull-right",
```
Comment (KatherineMuedas) [inline app/views/imports/_volunteers.html.erb:26]: same here `:class =>` to `class:`

**Comment by KatherineMuedas** [comment]
Great call on changing the dialogs!! 🚀

File: app/controllers/imports_controller.rb [hunk 4]
```diff
@@ -1,30 +1,74 @@
...
+  ERR_FILE_NOT_FOUND = "CSV import file not found."
+  ERR_FILE_EMPTY = "File can not be empty."
+  ERR_INVALID_HEADER = "Looks like this CSV contains invalid formatting. " \
+    "Please download an example CSV for reference and try again."
+
+  def header
+    {
+      "volunteer" => ["display_name", "email"],
+      "supervisor" => ["email", "display_name", "supervisor_volunteers"],
+      "casa_case" => ["case_number", "transition_aged_youth", "case_assignment", "birth_month_year_youth"]
```
Comment (ds29) [inline app/controllers/imports_controller.rb:54]: We could probably wrap these arrays into constants within the individual classes. e.g. `VolunteerImporter::IMPORT_HEADERS` and then reuse those in the class itself (and tests).

Comment (ryanmrodriguez) [reply app/controllers/imports_controller.rb]: bab1e0f5a7ab9cfe4d0733799cf466fae32c972a
Also added a few specs for the header validation

**Review by KatherineMuedas — APPROVED** [review]
LGTM!

**Pull request merged**